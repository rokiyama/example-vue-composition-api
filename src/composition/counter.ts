import { ref } from "@vue/composition-api";

export function useCount() {
  const counter = ref(0);
  const increment = () => {
    counter.value++;
  };
  const decrement = () => {
    counter.value--;
  };
  return {
    counter,
    increment,
    decrement
  };
}
