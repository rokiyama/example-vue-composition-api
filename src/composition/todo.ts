import { reactive, computed, toRefs } from "@vue/composition-api";

export function useTodo() {
  const state = reactive<{ todos: string[] }>({
    todos: []
  });
  const length = computed(() => state.todos.length);
  const addTodo = (item: string) => {
    state.todos.push(item);
  };
  const deleteTodo = (index: number) => {
    state.todos.splice(index, 1);
  };
  return {
    ...toRefs(state),
    length,
    addTodo,
    deleteTodo
  };
}
