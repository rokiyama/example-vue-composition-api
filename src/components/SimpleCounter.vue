<template>
  <div class="hello">
    <h1>{{ counter }}</h1>
    <div class="button-box">
      <button class="plus" @click="increment">+</button>
      <button class="minus" @click="decrement">-</button>
    </div>
    {{ mode }}
  </div>
</template>

<script lang="ts">
import { createComponent } from "@vue/composition-api";

import { useCount } from "@/composition/counter";

export default createComponent({
  setup() {
    return {
      ...useCount(),
      mode: process.env.NODE_ENV
    };
  }
});
</script>
