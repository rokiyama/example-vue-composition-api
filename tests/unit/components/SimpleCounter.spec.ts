import VueCompositionApi, { ref } from "@vue/composition-api";
import { shallowMount, createLocalVue } from "@vue/test-utils";
import SimpleCounter from "@/components/SimpleCounter.vue";
import * as composition from "@/composition/counter";

const localVue = createLocalVue();
localVue.use(VueCompositionApi);

let incrementMock: jest.Mock;
let decrementMock: jest.Mock;

describe("SimpleCounter.vue", () => {
  beforeEach(() => {
    jest.mock("@/composition/counter");
    incrementMock = jest.fn();
    decrementMock = jest.fn();
    jest.spyOn(composition, "useCount").mockReturnValue({
      counter: ref(0),
      increment: incrementMock,
      decrement: decrementMock
    });
  });
  it("renders initial html", () => {
    const wrapper = shallowMount(SimpleCounter, {
      localVue
    });
    wrapper.find("button.plus").trigger("click");
    expect(incrementMock).toHaveBeenCalled();
  });
  it("call increment when plus buttons is clicked", () => {
    const wrapper = shallowMount(SimpleCounter, {
      localVue
    });
    wrapper.find("button.plus").trigger("click");
    expect(incrementMock).toHaveBeenCalled();
  });
  it("call increment when minus buttons is clicked", () => {
    const wrapper = shallowMount(SimpleCounter, {
      localVue
    });
    wrapper.find("button.minus").trigger("click");
    expect(decrementMock).toHaveBeenCalled();
  });
});
