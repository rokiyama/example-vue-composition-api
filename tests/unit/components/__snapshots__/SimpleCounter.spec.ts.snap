// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleCounter.vue renders initial html 1`] = `
<div class="hello">
  <h1>0</h1>
  <div class="button-box"><button class="plus">+</button> <button class="minus">-</button></div>
  test
</div>
`;
